### PR TITLE
[WIP] epic/kibana packages cmd

### PR DESCRIPTION
**Pull request by JaminB**

### CODE REVIEW

File: dynamite_nsm/services/kibana/package/manager.py [hunk 1]
```diff
@@ -0,0 +1,784 @@
...
+        Args:
+            query (dict): ES DSL for the query
+
+        Raises:
+            ValueError: If response from ES is nonsuccess (not 200 series status)
+
+        Returns:
+            result: dict
+        """
+        result = requests.get(f"{Package.es_url}/{Package.package_index_name}/_search/",
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:105]: There's a very good possiblity that a `requests.exceptions.ConnectionError` will be thrown here.

Please log the error, and then raise the exception

File: dynamite_nsm/services/kibana/package/manager.py [hunk 2]
```diff
@@ -0,0 +1,784 @@
...
+            self.data['slug'] = self.slug
+        return json.dumps(self.data)
+
+    def __repr__(self) -> str:
+        return f"<PackageManifest(name={self.name}, author={self.author})>"
+
+class Package(object):
+
+    package_index_name = PACKAGES_INDEX_NAME
+    es_url = f'https://{get_primary_ip_address()}:9200'
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:77]: We can't always assume this. Totally possible that ES doesn't live local to the device the package manager lives on. We also cannot assume the port number. This is also configurable.

Considering kibana package manager can be used as a standalone application we cannot even always rely on retreiving the right target from the configuration files.


Here's the logic I'd apply to the situation.

Check if Kibana is installed locally. (kibana.profile.ProfileManager().is_installed()).
- if it is installed, grab the location of the host/port via the`kibana.install.ConfigManager().elasticsearch_targets` (Use index 0).
- If it is not installed, force the user to provide the target-string as an argument.

File: dynamite_nsm/services/kibana/package/manager.py [hunk 3]
```diff
@@ -0,0 +1,784 @@
...
+                    if sel-1 not in range(0, len(installed_packages)):
+                        raise ValueError(INVALID_SELECTION)
+                    selections.append(sel)
+            except ValueError as e:
+                if str(e) == INVALID_SELECTION:
+                    numpkgs = len(installed_packages)
+                    rangemsg = ""
+                    if numpkgs > 1:
+                        rangemsg = f". Must be 1-{numpkgs}"
+                    print(f"{INVALID_SELECTION}{rangemsg}")
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:517]: Please use `self.logger` instead of print

File: dynamite_nsm/services/kibana/package/manager.py [hunk 4]
```diff
@@ -0,0 +1,784 @@
...
+            Package: a Package instance from the loaded package file.
+        """
+        tar = tarfile.open(package_path)
+        try:
+            manifest = tar.extractfile('manifest.json')
+            manifest = PackageManifest(manifest.read().decode('utf8'))
+            package = Package(manifest)
+            return package
+        except KeyError:
+            print("Package must contain a manifest.json")
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:188]: Please log the error then raise the exception. Do not use print except maybe for user input prompts.

File: dynamite_nsm/services/kibana/package/manager.py [hunk 5]
```diff
@@ -0,0 +1,784 @@
...
+        packagesdata = [r['_source'] for r in result['hits']['hits']]
+        packages = []
+        for pkg in packagesdata:
+            manifest = PackageManifest(pkg.get('manifest'))
+            instobjs = [InstalledObject.from_kwargs(**iobj) for iobj in pkg.get('installed_objects')]
+            package = Package(manifest, instobjs, id=pkg.get('id'))
+            packages.append(package)
+        return packages
+
+class SavedObjectsManager(object):
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:426]: The use of `object` is depreciated in Python3 as everything is derived by object by default now.

File: dynamite_nsm/services/kibana/package/manager.py [hunk 6]
```diff
@@ -0,0 +1,784 @@
...
+    def __init__(self,
+                 stdout: Optional[bool] = True,
+                 verbose: Optional[bool] = False):
+        """Initializes the SavedObjectsManager
+
+        Args:
+            stdout (Optional[bool], optional): [description]. Defaults to True.
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:433]: Please use the following for stdout/verbose to maintain consistency across other command-line utilities.       
```      
stdout: Print the output to console
verbose: Include detailed debug messages
```

File: dynamite_nsm/services/kibana/package/manager.py [hunk 7]
```diff
@@ -0,0 +1,784 @@
...
+            print(f"Found {len(filepaths)} packages to install.")
+        for _filepath in filepaths:
+            filepath = os.path.abspath(_filepath)
+            # check mimetype of the file to determine how to proceed
+            filetype, encoding = mimetypes.MimeTypes().guess_type(filepath)
+            installation_statuses = []
+            # default to orphan package in case of install from .ndjson
+            manifest = PackageManifest(OrphanPackageData)
+            package = Package(manifest)
+            print(filepath)
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:638]: If you still need this info, please add as a `logger.debug` message

File: dynamite_nsm/services/kibana/package/manager.py [hunk 8]
```diff
@@ -0,0 +1,784 @@
...
+        else:
+            # check for installable items by extension, they will be verified by mimetype later.
+            acceptable_extensions = ['tar.xz', 'tar.gz']
+            for itm in os.listdir(package_install_path):
+                for ex in acceptable_extensions:
+                    if itm.endswith(ex):
+                        filepaths.append(f"{package_install_path}{itm}")
+                        break
+            print(f"Found {len(filepaths)} packages to install.")
+        for _filepath in filepaths:
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:630]: Might make sense to break out the conditionals in this for loop into separate sub-functions. It's a bit hard to follow.

File: dynamite_nsm/services/kibana/package/manager.py [hunk 9]
```diff
@@ -0,0 +1,784 @@
...
+                tar = tarfile.open(filepath)
+                try:
+                    manifest = tar.extractfile('manifest.json')
+                    manifest = PackageManifest(manifest.read().decode('utf8'))
+                    package = Package(manifest)
+                    package.create_packages_index()
+                    # check if package exists already.
+                    existing = Package.find_by_slug(package.manifest.create_slug())
+                    if existing:
+                        rmexisting = input(f"A Package titled {existing.manifest.name} is already installed, do you want to uninstall it? [y/n]") in "yY"
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:650]: The user should be able to bypass this prompt. Please provide a boolean in this method that can be converted to commandline flag (E.G ignore_warnings).

File: dynamite_nsm/services/kibana/package/manager.py [hunk 10]
```diff
@@ -0,0 +1,784 @@
...
+                        if rmexisting:
+                            rmsuccess = existing.uninstall(self.kibana_url, auth=(username, password), force=True)
+                            if rmsuccess:
+                                print(f"Successfully removed existing package {existing.manifest.name}.")
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:654]: self.logger.info

File: dynamite_nsm/services/kibana/package/manager.py [hunk 11]
```diff
@@ -0,0 +1,784 @@
...
+                    
+                except KeyError:
+                    print("Package must contain a manifest.json")
+                    exit(0)
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:658]: log the error raise the exception

File: dynamite_nsm/services/kibana/package/manager.py [hunk 12]
```diff
@@ -0,0 +1,784 @@
...
+                    kibana_objects_file.close()
+                    installation_statuses.append(self._process_package_installation_results(package, result))
+            # Should we remove this and ONLY install validatable packages?
+            elif filetype in ('application/json', 'text/plain') and filepath.endswith('ndjson'):
+                with open(filepath, 'r') as kibana_objects_file:
+                    result = self.import_kibana_saved_objects(username=username, password=password,
+                                                    kibana_objects_file=kibana_objects_file)
+                    installation_statuses.append(self._process_package_installation_results(package, result))
+            else:
+                self.logger.error("Files must be one of: .ndjson, .json, .tar.xz, .tar.gz")
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:673]: Line `623` implies that the only `acceptable_extensions` are ['tar.xz', 'tar.gz']

File: dynamite_nsm/services/kibana/package/manager.py [hunk 13]
```diff
@@ -0,0 +1,784 @@
...
+    def list(self, username: Optional[str] = None, password: Optional[str] = None, pretty: Optional[bool] = False):
+        """List packages currently installed for this instance
+
+        Args:
+            username (Optional[str], optional): kibana auth username. Defaults to None.
+            password (Optional[str], optional): kibana auth passwd. Defaults to None.
+        """
+        packages = Package.search_installed_packages()
+        if not packages:
+            print("Could not find any installed packages")
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:693]: logger.error

File: dynamite_nsm/services/kibana/package/manager.py [hunk 14]
```diff
@@ -0,0 +1,784 @@
...
+            # TODO raise exception
+        return resp.json()
+
+    def install(self, package_install_path: str, username: Optional[str] = None, password: Optional[str] = None):
+        """Install a package. A package can be given as an archive or directory. A package must contain one or more ndjson files and a manifest.json
+
+        Args:
+            package_install_path (str): path to the file or folder of files
+            username (Optional[str], optional): kibana auth usrname. Defaults to None.
+            password (Optional[str], optional): kibana auth password. Defaults to None.
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:615]: Default username and password to `admin` `admin` instead of `None`.

Otherwise,  the following calls will fail or require exception handling down the line.

```python3
652: rmsuccess = existing.uninstall(self.kibana_url, auth=(username, password), force=True)
```

```python3
662:  result = self.import_kibana_saved_objects(username=username, password=password,
663:                                                 kibana_objects_file=kibana_objects_file)
```

File: dynamite_nsm/services/kibana/package/manager.py [hunk 15]
```diff
@@ -0,0 +1,784 @@
...
+            else:
+                self.logger.error("Files must be one of: .ndjson, .json, .tar.xz, .tar.gz")
+                # TODO raise exception
+        
+        if not all(installation_statuses):
+            print(f"\r\n{package.manifest.name} installation failed.")
+            if not self.verbose:
+                print("Use --verbose flag to see more error detail.\n")
+        else:
+            package.register()
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:681]: `package` is referenced potentially before assignment (if `filelist` is empty)

File: dynamite_nsm/services/kibana/package/manager.py [hunk 16]
```diff
@@ -0,0 +1,784 @@
...
+                                                    kibana_objects_file=kibana_objects_file)
+                    installation_statuses.append(self._process_package_installation_results(package, result))
+            else:
+                self.logger.error("Files must be one of: .ndjson, .json, .tar.xz, .tar.gz")
+                # TODO raise exception
+        
+        if not all(installation_statuses):
+            print(f"\r\n{package.manifest.name} installation failed.")
+            if not self.verbose:
+                print("Use --verbose flag to see more error detail.\n")
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:679]: Please use logger

File: dynamite_nsm/services/kibana/package/manager.py [hunk 17]
```diff
@@ -0,0 +1,784 @@
...
+                with open(filepath, 'r') as kibana_objects_file:
+                    result = self.import_kibana_saved_objects(username=username, password=password,
+                                                    kibana_objects_file=kibana_objects_file)
+                    installation_statuses.append(self._process_package_installation_results(package, result))
+            else:
+                self.logger.error("Files must be one of: .ndjson, .json, .tar.xz, .tar.gz")
+                # TODO raise exception
+        
+        if not all(installation_statuses):
+            print(f"\r\n{package.manifest.name} installation failed.")
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:677]: Please use logger.error

File: dynamite_nsm/services/kibana/package/manager.py [hunk 18]
```diff
@@ -0,0 +1,784 @@
...
+                row.append(total)
+                table.append(row)
+            print(tabulate(table, headers=headers, tablefmt="fancy_grid"))
+        else:
+            data = []
+            for package in packages:
+                data.append(package.es_input())
+            print(data)
+        
+    def list_saved_objects(self, username: Optional[str] = None, password: Optional[str] = None,
```
Comment (JaminB) [inline dynamite_nsm/services/kibana/package/manager.py:721]: Let's return a string here instead of printing it, defaults to json. The cmd modules enable this functionality automatically for `entry_methods` in single/multi resp interfaces

Here's a reference implementation:  https://github.com/DynamiteAI/dynamite-nsm/blob/feature/configmanager-tests/dynamite_nsm/services/base/process.py#L100-L166

File: dynamite_nsm/services/kibana/package/manager.py (hunk 12 above)
Comment (Cameleopardus) [reply dynamite_nsm/services/kibana/package/manager.py:673]: Resolved:
>  installing from an array of ndjson files could lead to some issues down the road since the only way to uninstall them would be to uninstall the entire orphaned package.
Is the reason for this, we can point directly to a file to install it but if installing many from a folder, only valid packages are picked up.

File: dynamite_nsm/services/kibana/package/manager.py (hunk 15 above)
Comment (louism-dynamiteai) [reply dynamite_nsm/services/kibana/package/manager.py:681]: filelist is never empty, it is validated as a list with at least one item from the schema.

File: dynamite_nsm/services/kibana/package/manager.py (hunk 14 above)
Comment (Cameleopardus) [reply dynamite_nsm/services/kibana/package/manager.py:615]: I think this was the bug that happened during my demo.

File: dynamite_nsm/services/kibana/package/manager.py (hunk 2 above)
Comment (louism-dynamiteai) [reply dynamite_nsm/services/kibana/package/manager.py]: Update made to use kibana API to proxy request to ES

File: dynamite_nsm/services/kibana/package/manager.py (hunk 12 above)
Comment (louism-dynamiteai) [reply dynamite_nsm/services/kibana/package/manager.py]: Only valid extensions for multiple packages installed at once (from folder) are tar.gz/tar.xz. One-offs must be directly to the ndjson file.

File: dynamite_nsm/services/kibana/package/manager.py (hunk 2 above)
Comment (louism-dynamiteai) [reply dynamite_nsm/services/kibana/package/manager.py]: Further changes made per discussion during scrum. Now provides a `target` option for the user to supply the url to kibana (validated with urlparser), as well as preferring `host:port` from kibana config data pulled from dynamite environment file, and finally falling back to `primary_ip:5601` if all else fails.
